File: P1/P1S1E2/lib/ProductoBicicleta.rb
public
class ProductoBicicleta

	# @ReturnType int
	public
	def getID()
		return @iD
	end

	# @ParamType iD int
	# @ReturnType void
	public
	def setID(iD)
		@iD = iD
	end

	# @ReturnType String
	public
	def toString()
		return "Bicicleta " + @iD.to_s
	end

	# @ReturnType ProductoBicicleta
	public
	def clone()
    return Marshal.load(Marshal.dump(self)) #Copia Profunda
	end
  
  public
  def comienzaCarrera(segundoRetirada)
    for i in (0..segundoRetirada - 1)
      puts toString + " en el segundo " + i.to_s + "\n"
      
      sleep(1) #Duerme 1 segundo
    end
    
    puts toString + " se retira en el segundo " + segundoRetirada.to_s + "\n"
  end

	public
	def initialize()
    if(self.class == ProductoBicicleta) #Si se está intentando instanciar ProductoBicicleta
      raise "ProductoBicicleta no debe ser instanciado"
    else
      # @AttributeType int
      @iD
    end
	end
end

File: P1/P1S1E1/src/P1S1E1/ProductoBicicleta.java
package P1S1E1;

public abstract class ProductoBicicleta  extends Thread{

	private int identificador;
        private int segundoRetirada;

	public void setIdentificador(int id) {
		this.identificador = id;
	}
                
        public int getID(){
            return this.identificador;
        }
        
        public void setSegundoRetirada(int segundo) {
		this.segundoRetirada = segundo;
	}
                
        public int getSegundoRetirada() {
		return this.segundoRetirada;
	}

        @Override
        public String toString(){
            return "Bicicleta " + this.identificador;
        }

        @Override
        public void run(){
            for (int i=0; i<segundoRetirada; i++){
                System.out.println(this + " en el segundo " + i);
                    try {
                        this.sleep(1000); //Duerme 1 segundo
                    } catch (Exception e) {
                        System.out.println(e);
                    }
            }
            System.out.println(this+" se retira en el segundo " + segundoRetirada);
        }
}